app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/claims/EnterClaimController.scala

```diff
@@ -52,609 +49,178 @@ class EnterClaimController @Inject() (
-
-        val result = EitherT
-          .liftF(updateSession(sessionStore, request)(_.copy(journeyStatus = Some(updatedJourney))))
-          .leftMap((_: Unit) => Error("could not update session"))
-
-        result.fold(
-          e => {
-            logger.warn("could not complete claim state", e)
+      withAnswers[ClaimsAnswer] { (fillingOutClaim, _) =>
+        fillingOutClaim.draftClaim
+          .fold(_.movementReferenceNumber)
+          .fold {
+            logger.warn("could not find movement reference number")
             errorHandler.errorResult()
-          },
-          _ =>
-            fillingOutClaim.draftClaim
-              .fold(_.movementReferenceNumber)
-              .fold(Redirect(routes.BankAccountController.enterBankAccountDetails()))(_ =>
-                Redirect(routes.BankAccountController.checkBankAccountDetails())
-              )
-        )
-
+          }(referenceNumber =>
+            referenceNumber.value match {
+              case Left(_)  => Redirect(routes.BankAccountController.enterBankAccountDetails())
+              case Right(_) => Redirect(routes.BankAccountController.checkBankAccountDetails())
+            }
+          )
       }
     }
 
+  protected def calculateNextPage(claimAnswer: ClaimsAnswer): Result =
+    claimAnswer.find(_.isFilled === false) match {
+      case Some(claim) =>
+        Redirect(routes.EnterClaimController.enterClaim(claim.id))
+      case None        =>
+        Redirect(routes.EnterClaimController.checkClaim())
+    }
+
+  protected def updateClaimAnswer(claimAnswer: ClaimsAnswer, fillingOutClaim: FillingOutClaim, nextPage: Result)(
+    implicit request: RequestWithSessionData[AnyContent]
+  ): Future[Result] = {
+    val newDraftClaim  = fillingOutClaim.draftClaim.fold(_.copy(claimsAnswer = Some(claimAnswer)))
+    val updatedJourney = fillingOutClaim.copy(draftClaim = newDraftClaim)
+    EitherT
+      .liftF(updateSession(sessionStore, request)(_.copy(journeyStatus = Some(updatedJourney))))
+      .leftMap((_: Unit) => Error("could not update session"))
+      .fold(
+        e => {
+          logger.warn("could not save claims", e)
+          errorHandler.errorResult()
+        },
+        _ => nextPage
+      )
+  }
+
+  protected def replaceUpdateRedirect(claimAnswer: ClaimsAnswer, newClaim: Claim, fillingOutClaim: FillingOutClaim)(
+    implicit request: RequestWithSessionData[AnyContent]
+  ): Future[Result] = {
+    val notEditedClaims = claimAnswer.toList.filterNot(_.id === newClaim.id)
+    val claims          = ClaimsAnswer(newClaim, notEditedClaims: _*)
+    val nextPage        = calculateNextPage(claims)
+    updateClaimAnswer(claims, fillingOutClaim, nextPage)
+  }
+
 }
 
 object EnterClaimController {
 
-  final case class AmountPair(
-    paidAmount: BigDecimal,
-    claimAmount: BigDecimal
-  )
-
   final case class ClaimAmount(amount: BigDecimal)
 
+  final case class ClaimAndPaidAmount(paidAmount: BigDecimal, claimAmount: BigDecimal)
```

Comment:
Is it possible to collapse these two types into one which would handle both the mrn and entry number flow?

Reply:
Maybe it is possible, however, I don't think it's a good idea. One is used for the MRN journey from, the other on the entry number form, when you have to provide both.